File: MD/Fortran/MD.f
C
C  Simple molecular dynamics code.
C  $Id: MD.f,v 1.2 2002/01/31 16:43:14 spb Exp spb $
C
C This program implements:
C     long range inverse square forces between particles. F = G * m1*m2 / r**2
C     viscosity term     F = -u V
C If 2 particles approach closer than Size we flip the direction of the
C interaction force to approximate a collision.
C
C
C This program was developed as part of a code optimisation course
C and is therefore deliberately inefficient.
C
      SUBROUTINE evolve( count, dt)
      IMPLICIT NONE
      INCLUDE 'coord.inc'
      INTEGER count, step
      DOUBLE PRECISION dt
      DOUBLE PRECISION force
      INTEGER i,j,k,l

C
C Loop over timesteps.
C
      DO step = 1,count
        write(*,*) 'timestep ',step
        write(*,*) 'collisions ',collisions

C set the viscosity term in the force calculation
        DO j=1,Ndim
          CALL visc_force(Nbody,f(1,j),visc,vel(1,j))
        END DO

C calculate distance from central mass
        DO k=1,Nbody
          r(k) = 0.0
        END DO
        DO i=1,Ndim
          call add_norm(Nbody,r,pos(1,i))
        END DO
        DO k=1,Nbody
          r(k) = SQRT(r(k))
        END DO

C calculate central force

        DO i=1,Nbody
          DO l=1,Ndim
                f(i,l) = f(i,l) - 
     $              force(G*mass(i)*M_central,pos(i,l),r(i))
          END DO
        END DO

C calculate pairwise separation of particles
        k = 1
        DO i=1,Nbody
          DO j=i+1,Nbody
            DO l=1,Ndim
              delta_x(k,l) = pos(i,l) - pos(j,l)
            END DO
            k = k + 1
          END DO
        END DO

C calculate norm of seperation vector
        DO k=1,Npair
          delta_r(k) = 0.0
        END DO
        DO i=1,Ndim
          call add_norm(Npair,delta_r,delta_x(1,i))
        END DO
        DO k=1,Npair
          delta_r(k) = SQRT(delta_r(k))
        END DO

C
C add pairwise forces.
C
        k = 1
        DO i=1,Nbody
          DO j=i+1,Nbody
            DO l=1,Ndim
C  flip force if close in
              IF( delta_r(k) .GE. Size ) THEN
                f(i,l) = f(i,l) - 
     $              force(G*mass(i)*mass(j),delta_x(k,l),delta_r(k))
                f(j,l) = f(j,l) +
     $              force(G*mass(i)*mass(j),delta_x(k,l),delta_r(k))
              ELSE
                f(i,l) = f(i,l) +
     $               force(G*mass(i)*mass(j),delta_x(k,l),delta_r(k))
                f(j,l) = f(j,l) -
     $               force(G*mass(i)*mass(j),delta_x(k,l),delta_r(k))
                collisions = collisions + 1
              END IF
            END DO
            k = k + 1
          END DO
        END DO

C update positions
        DO i=1,Nbody
          DO j=1,Ndim
            pos(i,j) = pos(i,j) + dt * vel(i,j)
          END DO
        END DO

C update velocities
        DO i=1,Nbody
          DO j=1,Ndim
            vel(i,j) = vel(i,j) + dt * (f(i,j)/mass(i))
          END DO
        END DO


      END DO

      END

File: MD/C/MD.c
/*
 *  Simple molecular dynamics code.
 *  $Id: MD-c.c,v 1.2 2002/01/31 16:43:14 spb Exp spb $
 *
 * This program implements:
 *     long range inverse square forces between particles. F = G * m1*m2 / r**2
 *     viscosity term     F = -u V
 * If 2 particles approach closer than Size we flip the direction of the
 * interaction force to approximate a collision.
 *
 *
 * This program was developed as part of a code optimisation course
 * and is therefore deliberately inefficient.
 *
 */
#include <math.h>
#include "coord.h"

void visc_force(int N,double *f, double *visc, double *vel);
void add_norm(int N,double *r, double *delta);
double force(double W, double delta, double r);






void evolve(int count,double dt){
int  step;
int i,j,k,l;

/*
 * Loop over timesteps.
 */
      for(step = 1;step<=count;step++){
        printf("timestep %d\n",step);
        printf("collisions %d\n",collisions);

/* set the viscosity term in the force calculation */
        for(j=0;j<Ndim;j++){
          visc_force(Nbody,f[j],visc,vel[j]);
        }
/* calculate norm of seperation vector */
        for(k=0;k<Nbody;k++){
          r[k] = 0.0;
        }
        for(i=0;i<Ndim;i++){
	  add_norm(Nbody,r,pos[i]);
        }
        for(k=0;k<Nbody;k++){
          r[k] = sqrt(r[k]);
        }
       /* calculate central force */
        for(i=0;i<Nbody;i++){
	  for(l=0;l<Ndim;l++){
                f[l][i] = f[l][i] - 
                   force(G*mass[i]*M_central,pos[l][i],r[i]);
	  }
	}
/* calculate pairwise separation of particles */
        k = 0;
        for(i=0;i<Nbody;i++){
          for(j=i+1;j<Nbody;j++){
            for(l=0;l<Ndim;l++){
              delta_x[l][k] = pos[l][i] - pos[l][j];
            }
            k = k + 1;
          }
        }

/* calculate norm of seperation vector */
        for(k=0;k<Npair;k++){
          delta_r[k] = 0.0;
        }
        for(i=0;i<Ndim;i++){
	  add_norm(Npair,delta_r,delta_x[i]);
        }
        for(k=0;k<Npair;k++){
          delta_r[k] = sqrt(delta_r[k]);
        }

/*
 * add pairwise forces.
 */
        k = 0;
        for(i=0;i<Nbody;i++){
          for(j=i+1;j<Nbody;j++){
            for(l=0;l<Ndim;l++){
/*  flip force if close in */
              if( delta_r[k] >= Size ){
                f[l][i] = f[l][i] - 
                   force(G*mass[i]*mass[j],delta_x[l][k],delta_r[k]);
                f[l][j] = f[l][j] + 
                   force(G*mass[i]*mass[j],delta_x[l][k],delta_r[k]);
              }else{
                f[l][i] = f[l][i] + 
                   force(G*mass[i]*mass[j],delta_x[l][k],delta_r[k]);
                f[l][j] = f[l][j] - 
                   force(G*mass[i]*mass[j],delta_x[l][k],delta_r[k]);
		collisions++;
              }
            }
            k = k + 1;
          }
        }

/* update positions */
        for(i=0;i<Nbody;i++){
          for(j=0;j<Ndim;j++){
            pos[j][i] = pos[j][i] + dt * vel[j][i];
          }
        }

/* update velocities */
        for(i=0;i<Nbody;i++){
          for(j=0;j<Ndim;j++){
            vel[j][i] = vel[j][i] + dt * (f[j][i]/mass[i]);
          }
        }


      }

}




